File: c/matching_brackets.c
#include "matching_brackets.h"

#include <stddef.h>
#include <stdlib.h>
#include <string.h>

bool is_paired(const char *input) {
  char *stack = malloc(strlen(input) + 16);
  size_t top = 0;
  while (*input != '\0') {
    switch (*input) {
    case '(':
      stack[top++] = '(';
      break;
    case ')':
      if (top == 0 || stack[top - 1] != '(') {
        return false;
      }
      top--;
      break;
    case '[':
      stack[top++] = '[';
      break;
    case ']':
      if (top == 0 || stack[top - 1] != '[') {
        return false;
      }
      top--;
      break;
    case '{':
      stack[top++] = '{';
      break;
    case '}':
      if (top == 0 || stack[top - 1] != '{') {
        return false;
      }
      top--;
      break;
    default:
      break;
    }
    input++;
  }

  free(stack);
  return top == 0;
}

File: dart/matching_brackets.dart
class MatchingBrackets {
  bool isPaired(String s) {
    List<String> st = [];
    for (int i = 0; i < s.length; i++) {
      if ('([{'.contains(s[i])) {
        st.add(s[i]);
      } else if (s[i] == ')') {
        if (st.isEmpty || st.last != '(') {
          return false;
        }
        st.removeLast();
      } else if (s[i] == '}') {
        if (st.isEmpty || st.last != '{') {
          return false;
        }
        st.removeLast();
      } else if (s[i] == ']') {
        if (st.isEmpty || st.last != '[') {
          return false;
        }
        st.removeLast();
      }
    }

    return st.isEmpty;
  }
}
